rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,710 @@
+within the limitations of the Rust language itself, and what we're comfortable with representing via
+attributes. For example, single-line and multi-line doc comments are trivially supported, but if we
+wanted to start pushing more semantically-relevant information, such as configuration snippets, etc,
+it could be technically posssible but appear as a very ugly attribute usage, which could make the
+source muddy and unclear.
+
+Additionally, while derive macros are written in Rust code, which can be much easier to grok than
+declarative macros, it represents a section of code that may be harder to understand and modify than
+if we used a more brute force approach.
+
+## Prior Art
+
+Notably, and perhaps most obviously, Cue itself can be used for data validation purposes. There are
+also other projects that slot themselves into the same general domain, such as [OPA/Rego][opa_rego]
+and [CDDL][cddl].
+
+The primary problem with all of these solutions is that they're all custom languages, with far less
+cross-platform support, and a far steeper learning curve. Even though our approach generally
+concerns itself with obfuscating the schema tool itself, and focusing on making it trivially to
+generate the schema from annotated Rust types/fields, eventually the rubber must meet the road, and
+this is where these other tools would fall down for our case and become very hard to wield
+correctly.
+
+## Alternatives
+
+While `schemars` itself has support for annotating Rust types in almost the exact same way as we've
+proposed above, it lacks a few features necessary for our use case:
+
+- no support for the serde `alias` attribute feature
+- no support for defining generic metadata for a type/field and exposing it in the schema
+- no mechanism to override constraints for a field which already defines its own constraints at the
+  type level
+
+The lack of these features make it much harder to correctly generate our schema, as we would still
+be required to do the minimum amount of work to support defining custom attributes for the missing
+features, and the work to support at least one attribute is 90% of the work to support two, or
+three, or more, custom attributes. It would also mean that developers would need to figure out
+whether or not `schemars` provided a certain behavior via its attributes, versus being able to only
+need to remember how to get to the documentation for the our proposed internal approach.
+
+## Outstanding Questions
+
+- Can the proposed solution work seamlessly for all types that we use in our configuration? (For
+  example, `Option<T>` or any other type that needs to partially forward its declaration to the
+  underlying `T`)
+- Can downstream consumers extract the necessary information purely from the specified JSONSchema
+  fields and any additional custom metadata as string key/value pairs?  Do they need a richer
+  representation of custom metadata?
```

Comment:
I think string key/value pairs is fine for now. The only metadata I can think of that would violate this would be `min` and `max` values for a field, but I assume we'd represent those as part of JSON schema's [`maximum` attribute](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.2.2)?

Reply:
Yeah, anything like minimum/maximum, allowed patterns, required/optional, etc... are all representable with native JSONSchema keywords and need no custom metadata/extensions.